# 2 - YouTube RAG con GPT3.5 Turbo y LangChain


<br>
<br>

<img src="https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/rag_2.webp" style="width:400px;"/>

<h1>Tabla de Contenidos<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1.-Definiciones-y-dependencias." data-toc-modified-id="1.-Definiciones-y-dependencias.-0">1. Definiciones y dependencias.</a></span></li><li><span><a href="#2.-Importamos-la-API-KEY-de-OpenAI." data-toc-modified-id="2.-Importamos-la-API-KEY-de-OpenAI.-1">2. Importamos la API KEY de OpenAI.</a></span></li><li><span><a href="#3.-Llamada-a-GPT3.5-desde-LangChain" data-toc-modified-id="3.-Llamada-a-GPT3.5-desde-LangChain-2">3. Llamada a GPT3.5 desde LangChain</a></span><ul class="toc-item"><li><span><a href="#Output-Parser" data-toc-modified-id="Output-Parser-2.1"><a href="https://python.langchain.com/docs/modules/model_io/output_parsers/" rel="nofollow" target="_blank">Output Parser</a></a></span></li></ul></li><li><span><a href="#4.-Creación-de-la-cadena" data-toc-modified-id="4.-Creación-de-la-cadena-3">4. Creación de la cadena</a></span><ul class="toc-item"><li><span><a href="#LangChain-Expression-Language-(LCEL)" data-toc-modified-id="LangChain-Expression-Language-(LCEL)-3.1"><a href="https://python.langchain.com/docs/expression_language/" rel="nofollow" target="_blank">LangChain Expression Language (LCEL)</a></a></span></li></ul></li><li><span><a href="#5.-Plantillas-de-instrucciones-(Prompt-Templates)" data-toc-modified-id="5.-Plantillas-de-instrucciones-(Prompt-Templates)-4">5. Plantillas de instrucciones (<a href="https://python.langchain.com/docs/modules/model_io/prompts/quick_start/" rel="nofollow" target="_blank">Prompt Templates</a>)</a></span></li><li><span><a href="#6.-Combinando-cadenas" data-toc-modified-id="6.-Combinando-cadenas-5">6. Combinando cadenas</a></span></li><li><span><a href="#7.-Extracción-de-texto-desde-YouTube-con-Whisper" data-toc-modified-id="7.-Extracción-de-texto-desde-YouTube-con-Whisper-6">7. Extracción de texto desde YouTube con Whisper</a></span></li><li><span><a href="#8.-Usando-la-transcripción-como-contexto" data-toc-modified-id="8.-Usando-la-transcripción-como-contexto-7">8. Usando la transcripción como contexto</a></span><ul class="toc-item"><li><span><a href="#Dividiendo-en-segmentos-el-texto-completo" data-toc-modified-id="Dividiendo-en-segmentos-el-texto-completo-7.1">Dividiendo en segmentos el texto completo</a></span></li><li><span><a href="#Encontrando-los-trozos-relevantes" data-toc-modified-id="Encontrando-los-trozos-relevantes-7.2">Encontrando los trozos relevantes</a></span></li></ul></li><li><span><a href="#9.-Bases-de-datos-vectoriales" data-toc-modified-id="9.-Bases-de-datos-vectoriales-8">9. Bases de datos vectoriales</a></span><ul class="toc-item"><li><span><a href="#Conectando-la-base-de-datos-a-la-cadena" data-toc-modified-id="Conectando-la-base-de-datos-a-la-cadena-8.1">Conectando la base de datos a la cadena</a></span></li><li><span><a href="#Chroma" data-toc-modified-id="Chroma-8.2">Chroma</a></span></li></ul></li><li><span><a href="#10.-Resumen-paso-a-paso" data-toc-modified-id="10.-Resumen-paso-a-paso-9">10. Resumen paso a paso</a></span></li></ul></div>

## 1. Definiciones y dependencias. 

**Dependencias necesarias para este workshop:**

```bash
pip install langchain
pip install langchain-openai
pip install "langchain[docarray]"
pip install langchain-chroma
pip install docarray
pip install pydantic==1.10.9
pip install openai-whisper
pip install pytube
pip install scikit-learn
```

El propósito de este workshop es crear un RAG (Retrieval Augmented Generation) con [LangChain](https://www.langchain.com/) y el modelo GPT3.5 Turbo de [OpenAI](https://openai.com/).

**¿Qué es un LLM?**


Un LLM, o "Large Language Model" (Gran Modelo de Lenguaje), es un tipo de modelo de inteligencia artificial diseñado para entender, generar y trabajar con lenguaje humano a gran escala. Estos modelos se entrenan utilizando grandes cantidades de datos de texto para aprender patrones, estructuras del lenguaje y relaciones contextuales.

Características principales de los LLM:

1. Capacidad de Generación de Texto: Los LLMs son capaces de generar texto coherente y contextualmente relevante que puede imitar el estilo y la estructura del lenguaje humano. Esto los hace útiles para tareas como la escritura automática de artículos, generación de respuestas en chatbots, etc...

2. Comprensión del Contexto: Gracias a su entrenamiento con grandes cantidades de texto, los LLMs tienen una notable capacidad para entender el contexto de las consultas que reciben, lo que les permite ofrecer respuestas más precisas y relevantes.

3. Aplicaciones Multilingües: Algunos LLMs están entrenados en múltiples idiomas, lo que les permite operar en diferentes lenguajes y realizar tareas como traducción automática o asistencia multilingüe.

4. Aprendizaje Continuo: Aunque los LLMs se entrenan en un conjunto estático de datos, algunos modelos están diseñados para continuar aprendiendo a partir de nuevas interacciones, lo que mejora su rendimiento y adaptabilidad con el tiempo.

5. Interpretación de Sentimiento y Semántica: Pueden analizar y entender sentimientos, opiniones y matices semánticos en el texto, lo que es crucial para aplicaciones como análisis de sentimiento, soporte al cliente y monitorización de redes sociales.

Ejemplos de LLM:

+ GPT (Generative Pre-trained Transformer) de OpenAI: Es uno de los modelos de lenguaje más conocidos y avanzados, usado ampliamente en aplicaciones comerciales y de investigación por su capacidad para generar texto altamente contextual y creativo.

+ BERT (Bidirectional Encoder Representations from Transformers) de Google: Se utiliza principalmente para mejorar la comprensión del lenguaje en buscadores y para mejorar la precisión de las respuestas en aplicaciones de inteligencia artificial.


Cuando se habla de los parámetros de un LLM nos estamos refiriendo a los valores internos que el modelo utiliza para realizar predicciones y generar texto basado en el lenguaje humano. Estos parámetros son esenciales para que el modelo funcione correctamente y son ajustados durante el proceso de entrenamiento. Los parámetros son en su mayoría pesos sinápticos. Estos pesos determinan la fuerza de la conexión entre las neuronas en diferentes capas del modelo. Durante el entrenamiento, estos pesos se ajustan para minimizar el error en la predicción del modelo. Junto con los pesos, los biases (sesgos) son otro tipo de parámetros que se añaden a las sumas ponderadas en las neuronas para ayudar al modelo a ajustarse mejor a los datos, son el equivalente a la ordenada en el origen. Los biases permiten que el modelo opere eficazmente cuando todas las entradas son cero.

Los LLMs son conocidos por tener un número extremadamente alto de parámetros. Por ejemplo, GPT-3 de OpenAI tiene 175 mil millones de parámetros, lo que lo hace uno de los modelos de lenguaje más grandes y poderosos disponibles. Estos parámetros permiten al modelo capturar una gran cantidad de matices lingüísticos y contextuales.

**¿Qué es un RAG?**

La generación mejorada por recuperación (Retrieval Augmented Generation - RAG) es el proceso de optimización de un modelo lingüístico de gran tamaño (Large Language Model - LLM), de modo que conozca datos proporcionados por el usuario, que no existan en los datos de entrenamiento del modelo, antes de generar una respuesta. 

Los LLM se entrenan con grandes volúmenes de datos y usan miles de millones de parámetros para generar resultados originales en tareas como responder preguntas, traducir idiomas y completar frases. Un RAG extiende las ya poderosas capacidades de los LLM a dominios específicos o a la base de conocimientos interna de una organización, todo ello sin la necesidad de volver a entrenar el modelo. Se trata de un método rentable para mejorar los resultados de los LLM de modo que sigan siendo relevantes, precisos y útiles en diversos contextos.

Un RAG aporta varios beneficios directos en el desarrollo de una herramienta de inteligencia artificial:

+ Rentabilidad de la implementación. El desarrollo de IAs comienza normalmente con un modelo básico. Los modelos fundacionales (Foundational Models - FM) son LLMs accesibles por API entrenados en un amplio espectro de datos generalizados y sin etiquetar. Los costos computacionales y financieros de volver a entrenar a los FM para obtener información específica de la organización o del dominio son muy elevados. Un RAG es un enfoque más rentable para introducir nuevos datos en el LLM.


+ Información actualizada. Incluso si el LLM está entrenado con los datos adecuados para las necesidades de la compañía, es complicado mantener la relevancia del modelo. Un RAG permite a los desarrolladores proporcionar las últimas investigaciones, estadísticas o noticias a los modelos generativos. Se puede usar un RAG para conectar el LLM directamente a las redes sociales en vivo, sitios de noticias u otras fuentes de información que se actualizan con frecuencia. De esta manera, un LLM puede proporcionar la información más reciente.


+ Confianza. Al darle al LLM datos propios, se conoce perfectamente la fuente de datos además de evitar la alucinación del LLM.


+ Mayor control. El RAG permite a los desarrolladores de inteligencia artificial cambiar las fuentes de información para adaptarse a los requisitos cambiantes o a los usos múltiples de la compañía. Además pueden restringir la recuperación de información confidencial a diferentes niveles de autorización y garantizar que el LLM genere las respuestas adecuadas.


El esquema básico de un RAG es como sigue:

<br>

![rag](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/rag.png)

<br>

1. Los documentos de la compañía se pasan por un modelo de incrustación (Embedding Model) para ser guardados en forma de vectores en una base de datos.

2. La consulta realizada por el usuario pasa por el mismo modelo para convertir dicho texto en vectores.

3. Se buscan en la base de datos los vectores más parecidos a la consulta y se extraen los vectores más relevantes.

4. Los documentos más relevantes extraídos y la consulta se introducen en el LLM para generar la respuesta más adecuada.


![langchain](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/langchain.jpeg)

**¿Qué es LangChain?**


LangChain es un framework de código abierto para crear aplicaciones basadas en LLM. LangChain proporciona herramientas y abstracciones para mejorar la personalización, precisión y relevancia de la información que generan los modelos. Por ejemplo, los desarrolladores pueden usar los componentes de LangChain para crear nuevas cadenas de peticiones o personalizar las plantillas existentes. LangChain también incluye componentes que permiten a los LLM acceder a nuevos conjuntos de datos sin necesidad de repetir el entrenamiento.

![openai](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/openai.png)

**¿Qué es OpenAI?**

OpenAI es una organización dedicada a la investigación y aplicación de inteligencia artificial. Se centra en diversas áreas como modelos generativos, seguridad, robótica y procesamiento de lenguaje natural. Además, OpenAI proporciona una plataforma de API que permite a los usuarios crear y utilizar sus propios modelos de IA personalizados.



## 2. Importamos la API KEY de OpenAI. 

Para realizar este proyecto necesitaremos una API KEY de OpenAI para poder usar el modelo GPT3.5 Turbo. Dicha API KEY se obtiene en https://platform.openai.com/api-keys. Solamente se muestra una vez, por lo que hay que guardarla en el mismo momento en que se obtiene. Por supuesto, tendremos que crear una cuenta para obtenerla.

La API KEY la guardamos en un archivo `.env` para cargarla con la librería dotenv y usarla como variable de entorno. Dicho archivo se añade al `.gitignore` para que nadie pueda verla si subimos el código a GitHub por ejemplo.

In [1]:
# importamos la API KEY

import os                           # libreria del sistema operativo
from dotenv import load_dotenv      # carga variables de entorno 


load_dotenv()

OPENAI_API_KEY = os.getenv('OPENAI_API_KEY')

## 3. Llamada a GPT3.5 desde LangChain

A continuación, importamos el LLM de OpenAI desde LangChain. Aquí es donde usaremos nuestra API KEY para cargar el modelo GPT3.5 Turbo. Vamos a preguntarle quién ganó la liga española en 2020, por ejemplo, para comprobar que efectivamente funciona y que estamos correctamente conectados.

In [2]:
from langchain_openai.chat_models import ChatOpenAI   # conexión de LangChain a los modelos de OpenAI

modelo = ChatOpenAI(openai_api_key=OPENAI_API_KEY, model='gpt-3.5-turbo')

In [3]:
# invocamos al modelo para que genere la respuesta

respuesta = modelo.invoke('¿Quién ganó la liga española en 2020?')

In [4]:
# respuesta del modelo

respuesta

AIMessage(content='El Real Madrid ganó la liga española en 2020.', additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 15, 'prompt_tokens': 22, 'total_tokens': 37, 'completion_tokens_details': {'audio_tokens': None, 'reasoning_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': None, 'cached_tokens': 0}}, 'model_name': 'gpt-3.5-turbo-0125', 'system_fingerprint': None, 'finish_reason': 'stop', 'logprobs': None}, id='run-75f073ab-d1c2-4b83-bf56-693429627050-0', usage_metadata={'input_tokens': 22, 'output_tokens': 15, 'total_tokens': 37, 'input_token_details': {'cache_read': 0}, 'output_token_details': {'reasoning': 0}})

In [5]:
# tipo de dato de la respuesta

type(respuesta)

langchain_core.messages.ai.AIMessage

Como veis, la respuesta del modelo es un objeto de LangChain AIMessage. Si queremos extraer la string de la respuesta podemos hacerlo directamente con el atributo `content` de la respuesta. Probémoslo.

In [6]:
respuesta.content

'El Real Madrid ganó la liga española en 2020.'

### [Output Parser](https://python.langchain.com/docs/modules/model_io/output_parsers/)

El propio LangChain tiene un parser para la salida de string, `StrOutputParser`. Parse se refiere a cambiar el tipo de formato, básicamente es traducir. LangChain tiene muchos parsers incorporados, desde strings hasta csv pasando por JSON o incluso Pandas DataFrames. 

Desde el parser `StrOutputParser` de LangChain el modelo devuelve directamente la respuesta en formato string. Este parser lo usaremos a la hora de crear la cadena de LangChain, será un eslabón más de la misma y así obtendremos directamente la respuesta del LLM en formato string.

In [7]:
from langchain_core.output_parsers import StrOutputParser

parser = StrOutputParser()

## 4. Creación de la cadena

Una cadena de LangChain es una secuencia configurada de componentes o pasos que se encadenan para realizar tareas complejas de procesamiento del lenguaje. LangChain es una herramienta diseñada para facilitar la construcción y el despliegue de aplicaciones de lenguaje, integrando modelos de lenguaje avanzados y otras funcionalidades en flujos de trabajo coherentes y efectivos. El propósito de una cadena de LangChain es permitir que cada paso del proceso contribuya de manera efectiva al resultado final. Una cadena de LangChain típicamente incluirá varios componentes clave:

+ Recuperadores (Retrievers): Componentes que buscan y recuperan información relevante de una base de datos o de un conjunto de documentos. Estos son esenciales para proporcionar contexto o datos específicos que el modelo de lenguaje necesita para generar respuestas informadas.

+ Modelos de Lenguaje: El núcleo de una cadena de LangChain, donde se utilizan modelos avanzados como GPT-3 para generar texto, resolver preguntas, o realizar otras tareas de NLP basadas en la información y contextos proporcionados por otros componentes de la cadena.

+ Post-procesadores: Componentes que toman la salida del modelo de lenguaje y la refinan, por ejemplo, corrigiendo errores, formateando la respuesta, o aplicando filtros adicionales de contenido.




Ahora crearemos la cadena que tiene básicamente dos eslabones: el modelo y el parser a string. El flujo del proceso sería:

1. Generamos una consulta con una instrucción en texto (prompt).
2. Le pasamos la consulta al modelo que hemos inicializado.
3. El modelo devuelve una respuesta.
4. La respuesta del modelo entra en el parser.
5. El parser devuelve la respuesta en formato string.

![chain1](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/chain1.png)

In [8]:
# creamos la cadena

cadena = modelo | parser

In [9]:
# invocamos a la cadena dándole la consulta para que realice todo el flujo

cadena.invoke('¿Quién ganó la liga española en 2020?')

'El equipo que ganó la liga española en 2020 fue el Real Madrid.'

### [LangChain Expression Language (LCEL)](https://python.langchain.com/docs/expression_language/)

Fijaros en esta expresión:

```python
chain = model | parser
```

Esta expresión no es un [bitwise operator](https://realpython.com/python-bitwise-operators/) de Python, sino que usa para crear la cadena dentro de LangChain y concatenar las funciones dentro del diccionario del agente. Funcionaría como un grafo directo, haciendo que se ejecuten de una en una las tareas de la cadena.

## 5. Plantillas de instrucciones ([Prompt Templates](https://python.langchain.com/docs/modules/model_io/prompts/quick_start/))

Las plantillas de instrucciones son recetas predefinidas para generar instrucciones para modelos de lenguaje.

Una plantilla puede incluir instrucciones, contexto y preguntas específicas apropiadas para una tarea dada. LangChain proporciona herramientas para crear y trabajar con plantillas de instrucciones y además se esfuerza por crear plantillas agnósticas al modelo para facilitar la reutilización de plantillas existentes en diferentes modelos de lenguaje.

![chain2](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/chain2.png)

Típicamente, los modelos de lenguaje esperan que la instrucción sea una cadena de texto o una lista de mensajes de chat. Veamos un ejemplo:

In [10]:
from langchain.prompts import ChatPromptTemplate

In [11]:
# gpt3.5 funciona mejor en inglés, hagamos un ejemplo en esa lengua


# plantilla de texto con un contexto y una pregunta
plantilla = '''
            Answer the question based on the context below. If you can't 
            answer the question, reply "I don't know".

            Context: {context}

            Question: {question}
            '''

# carga de la plantilla en el prompt
prompt = ChatPromptTemplate.from_template(plantilla)


# formato de salida del prompt
prompt.format(context='The building where Pepe lives is green.', 
              question="What color is Pepe's house?").split('\n')

['Human: ',
 "            Answer the question based on the context below. If you can't ",
 '            answer the question, reply "I don\'t know".',
 '',
 '            Context: The building where Pepe lives is green.',
 '',
 "            Question: What color is Pepe's house?",
 '            ']

In [12]:
# creamos la cadena añadiendo la plantilla del prompt en el primer paso

cadena = prompt | modelo | parser

In [13]:
# invocamos a la cadena

cadena.invoke({'context': 'The building where Pepe lives is green.',
               'question': "What color is Pepe's house?"})

"The color of Pepe's house is green."

Fijaros que ahora en la invocación de la cadena no le pasamos una string, sino que le damos un diccionario que contiene tanto el contexto como la pregunta. Esto es necesario, puesto que el primer paso de la cadena es la creación del prompt y necesita de ambos términos para su creación.

In [14]:
# probemos directamente la cadena en castellano, a ver que ocurre

cadena.invoke({'context': 'Tengo un coche blanco',
               'question': '¿De qué color es mi coche?'})

'De color blanco.'

## 6. Combinando cadenas


Acabamos de ver que el modelo de lenguaje funciona también en castellano. Pero podemos crear cadenas para flujos de trabajo más complejos, como por ejemplo, traducir lo que hace la primera cadena a otro lenguaje. 

![chain3](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/chain3.png)

Usaremos otra plantilla de instrucciones para combinarla con la cadena que ya hemos construido, de tal manera que traduzca al castellano la respuesta del modelo. Veámoslo:

In [15]:
# creamos la plantilla de traducción con la respuesta de la cadena y el lenguaje de salida

prompt_traductor = ChatPromptTemplate.from_template('Translate {answer} to {language}')

In [16]:
# creamos la nueva cadena basada en la anterior a la cual le damos el lenguaje al que queremos traducir

cadena_traducida = (
    {'answer': cadena, 'language': lambda x: x['language']} | prompt_traductor | modelo | parser 
)

Fijaros que el primer paso de la cadena traducida es un diccionario con la respuesta de la cadena original por un lado y el lenguaje por el otro. La razón de la función `lambda` es que cuando invoquemos a esta cadena compuesta le pasaremos también un diccionario cuyos valores son ejecutables, como hicimos antes, con el contexto, la pregunta y el lenguaje al que queremos traducir la respuesta.

In [17]:
# invocamos a la nueva cadena, dándole el contexto, la pregunta y el lenguaje de salida

cadena_traducida.invoke({'context': 'The building where Pepe lives is green.',
                         'question': "What color is Pepe's house?",
                         'language': 'Spanish',
                         })

'La casa de Pepe es verde.'

In [18]:
# traducimos a otro idioma para probar la cadena

cadena_traducida.invoke({'context': 'The building where Pepe lives is green.',
                         'question': "What color is Pepe's house?",
                         'language': 'French',
                         })

'Vert'

In [19]:
# otro ejemplo

cadena_traducida.invoke({'context': 'The building where Pepe lives is green.',
                         'question': "What color is Pepe's house?",
                         'language': 'German',
                         })

"Pepe's Haus ist grün."

Hasta aquí, tenemos una cadena a la cual le damos un contexto, una pregunta y un lenguaje y nos devuelve una respuesta contextualizada en el idioma seleccionado. Nos hemos propuesto transcribir un video de YouTube para que la cadena lo use como contexto y podamos realizar consultas a dicha transcripción. Para ello lo primero será extraer el texto desde un video. Vamos a ello.

## 7. Extracción de texto desde YouTube con Whisper

Primero tenemos que importar las librerías que vamos a usar. Usaremos [pytube](https://pytube.io/en/latest/), para acceder al video y luego [whisper](https://openai.com/research/whisper), el modelo de speech-to-text de OpenAI. También tendremos que instalar [ffmpeg](https://ffmpeg.org/) según nuestro sistema operativo para que nuestra máquina sea capaz de analizar el audio que viene desde el video de YouTube.

In [20]:
from pytube import YouTube

import whisper

In [21]:
# definimos la url del video: Planned Chaos - by Ludwig von Mises - (Full Audiobook) (3:09:32)

VIDEO_URL = 'https://www.youtube.com/watch?v=7EnHeZXLzTc'

In [22]:
# usamos pytube para extraer el video

youtube = YouTube(VIDEO_URL)

In [23]:
# ahora extraemos el audio desde el video

audio = youtube.streams.filter(only_audio=True).first()

audio

<Stream: itag="140" mime_type="audio/mp4" abr="128kbps" acodec="mp4a.40.2" progressive="False" type="audio">

In [24]:
# se carga el modelo base de Whisper en local, 139M

modelo_whisper = whisper.load_model('base')

In [25]:
# descripción del modelo whisper

modelo_whisper

Whisper(
  (encoder): AudioEncoder(
    (conv1): Conv1d(80, 512, kernel_size=(3,), stride=(1,), padding=(1,))
    (conv2): Conv1d(512, 512, kernel_size=(3,), stride=(2,), padding=(1,))
    (blocks): ModuleList(
      (0-5): 6 x ResidualAttentionBlock(
        (attn): MultiHeadAttention(
          (query): Linear(in_features=512, out_features=512, bias=True)
          (key): Linear(in_features=512, out_features=512, bias=False)
          (value): Linear(in_features=512, out_features=512, bias=True)
          (out): Linear(in_features=512, out_features=512, bias=True)
        )
        (attn_ln): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (mlp): Sequential(
          (0): Linear(in_features=512, out_features=2048, bias=True)
          (1): GELU(approximate='none')
          (2): Linear(in_features=2048, out_features=512, bias=True)
        )
        (mlp_ln): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
      )
    )
    (ln_post): LayerNorm((512,), eps=1e-05,

**¿Cómo funciona whisper?**

Como veis, Whisper está básicamente dividido en dos partes: un AudioEncoder y un TextDecoder. Esto es lo que se conoce como AutoEncoder. El AudioEncoder reconoce el audio y lo vectoriza y desde esos vectores el TextDecoder saca el texto. Por daros un paso a paso sería algo así:
 

1. Preprocesamiento del Audio: Whisper comienza con el preprocesamiento del audio de entrada, donde se ajusta la calidad del audio para mejorar la precisión del reconocimiento. Esto puede incluir normalizar el volumen, filtrar ruido de fondo y segmentar el audio en trozos más manejables.

2. Extracción de Características: El modelo extrae características del audio procesado. Esto implica convertir las señales de audio en una forma que el modelo pueda entender, como espectrogramas o características de Mel-frequency cepstral coefficients (MFCC), que son representaciones de la frecuencia y amplitud del sonido a lo largo del tiempo.

3. Modelo de Aprendizaje: Whisper utiliza un modelo de red neuronal, entrenado en una gran cantidad de datos de audio y transcripciones correspondientes. Este modelo aprende a identificar patrones y correlaciones entre el audio y las transcripciones textuales.

4. Reconocimiento y Traducción: Durante la fase de reconocimiento, Whisper convierte el audio en texto utilizando su red neuronal. Además, puede traducir el texto reconocido a otros idiomas, gracias a su entrenamiento en múltiples idiomas.

5. Post-procesamiento: Finalmente, el texto generado pasa por un proceso de post-procesamiento para corregir errores comunes, ajustar la puntuación y mejorar la coherencia del texto.

Usaremos ahora Whisper para transcribir el audio a un archivo de texto. Primero crearemos un directorio temporal donde se guardará el audio y luego guardaremos la salida de Whisper en el archivo `transcripcion.txt`. Si ya existe el archivo, no nos interesa convertirlo otra vez, puesto que Whisper tarda unos minutos en trancribir todo el audio, asi que solo ejecutamos el modelo si el archivo de texto no existe.

In [26]:
import tempfile    # para manejo de archivos temporales


# ruta de guardado del archivo de texto

RUTA_TXT = 'txt/transcripcion.txt'

In [27]:
%%time

# si el archivo de texto no existe...
if not os.path.exists(RUTA_TXT):
    
    # abrimos el directorio temporal...
    with tempfile.TemporaryDirectory() as dir_temporal:
        
        # descargamos el audio de YouTube...
        archivo_audio = audio.download(output_path=dir_temporal)
        
        # y Whisper transcribe el audio a texto, en 32 bits (fp16=False)
        transcripcion = modelo_whisper.transcribe(archivo_audio, fp16=False)['text'].strip()
        
        
        # se guarda el archivo de texto
        with open(RUTA_TXT, 'w') as archivo_texto:
            archivo_texto.write(transcripcion)

CPU times: user 90 µs, sys: 64 µs, total: 154 µs
Wall time: 131 µs


In [28]:
# cargamos el archivo de texto para usarlo en la cadena

with open(RUTA_TXT, 'r') as archivo_texto:
    
    transcripcion = archivo_texto.read()

In [29]:
# 1000 primeros caracteres del texto
    
transcripcion[:1000]

'Planned Chaos by Ludwig von Meises Introductory remarks The characteristic mark of this age of dictators, wars and revolutions is its anti-capitalistic bias. Most governments and political parties are eager to restrict the sphere of private initiative and free enterprise. It is an almost unchallenged dogma that capitalism is done for, and that the coming of all-round regimentation of economic activities is both inescapable and highly desirable. Nonetheless, capitalism is still very vigorous in the Western Hemisphere. Capitalist production has made very remarkable progress even in these last years. Methods of production were greatly improved. Consumers have been supplied with better and cheaper goods and with many new articles unheard of a short time ago. Many countries have expanded the size and improved the quality of their manufacturing. In spite of the anti-capitalistic policies of all governments and of almost all political parties, the capitalist mode of production is in many cou

In [30]:
# nº de caracteres del texto

len(transcripcion)

159723

In [31]:
# nº de palabras del texto

len(transcripcion.split())

25880

In [32]:
# 1000 últimos caracteres

transcripcion[-1000:]

"n of a proletarian has contributed to the elaboration of the interventionist and socialist programmes. Their authors were all of bourgeois background. The esoteric writings of dialectical materialism, of Hegel, the father, both of Marxism and of German aggressive nationalism, the books of George's Sorrel, of Gentile and Spengler, are not read by the average man. They did not move the masses directly. It was the intellectuals who popularised them. The intellectual leaders of the peoples have produced and propagated the fallacies which are on the point of destroying liberty and western civilisation. The intellectuals alone are responsible for the mass slaughters which are the characteristic mark of our century. They alone can reverse the trend and pave the way for a resurrection of freedom. Not mythical material productive forces, but reason and ideas determine the course of human affairs. What is needed to stop the trend towards socialism and despotism is common sense and moral courage

## 8. Usando la transcripción como contexto

La idea fundamental del RAG es que el LLM tenga nuestra transcripción como contexto a la hora de generar respuestas, es decir, que conozca el texto para poder respondernos las preguntas que tengamos al respecto del mismo. Vamos a ver que ocurre cuando lo hacemos:

In [33]:
try:
    cadena.invoke({
        'context': transcripcion,
        'question': 'What are the key points of this book? Put them in a bullet point list.'})
    
except Exception as e:
    print(e)

Error code: 400 - {'error': {'message': "This model's maximum context length is 16385 tokens. However, your messages resulted in 30815 tokens. Please reduce the length of the messages.", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}


Si nos fijamos en el mensaje del error, nos está diciendo que la longitud máxima del contexto de GPT3.5 es de 16385 tokens pero nuestro contexto tiene 30815 tokens. Un token no es un caracter ni una palabra, recordad que nuestro texto tiene 159723 y 25880 respectivamente. Un token es una unidad de texto que el modelo utiliza para procesar y generar lenguaje. Los tokens son las piezas fundamentales de información con las que el modelo trabaja, y pueden variar en gran medida dependiendo de cómo está diseñado el modelo y cómo se ha entrenado. Un token puede ser una palabra completa, una parte de una palabra (como prefijos, raíces o sufijos), o incluso puntuación y espacios. La forma en que se dividen los textos en tokens depende del "tokenizador" utilizado por el modelo. La tokenización es uno de los procesos esenciales para el funcionamiento de los LLM, actuando como los bloques constructivos básicos para todas las tareas de procesamiento de lenguaje que realiza el modelo.

El problema aquí es que el texto es demasiado grande, de hecho el mensaje de error nos recomienda reducir la longitud del texto. Pero existe otra solución mejor a este problema que reducir el contenido del texto. Además, nosotros queremos que toda la transcripción sea el contexto que el LLM utilice para darnos una respuesta. Veamos esa solución.

### Dividiendo en segmentos el texto completo

![chain4](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/chain4.png)

Dado que el texto completo es demasiado grande para ser usado como contexto para el LLM, la solución obvia es dividir el texto en varios trozos para darle al modelo textos más pequeños como contexto. Para dividir el texto, primero usaremos el TextLoader de LangChain para cargar toda la transcripción del video como un objeto Document de LangChain. Un Document es un fragmento de texto con metadatos asociados. 

El cargador de texto proporciona un método "load" para cargar datos como Document desde una fuente como un txt o un md, por ejemplo. Opcionalmente, implementan también una "lazy load" para cargar los datos en la memoria. "Lazy" se refiere a una estrategia de carga o inicialización que pospone la ejecución de una operación hasta que es estrictamente necesaria. Esta técnica es comúnmente usada para mejorar el rendimiento y optimizar el uso de los recursos del sistema, especialmente la memoria.

In [34]:
from langchain_community.document_loaders import TextLoader

loader = TextLoader(RUTA_TXT)

documento_texto = loader.load()

In [35]:
len(documento_texto)

1

In [36]:
type(documento_texto[0])

langchain_core.documents.base.Document

Una vez cargado el texto completo como un objeto de LangChain, lo dividimos con un [text splitter](https://python.langchain.com/docs/modules/data_connection/document_transformers/). Existen distintos tipos de splitters según el tipo de archivo que queramos dividir, ya sea texto, HTML, markdown o código. Vamos a un usar un splitter recursivo. Dividir el texto de manera recursiva tiene el propósito de intentar mantener las piezas de texto relacionadas una con la siguiente. Es la manera recomendada de comenzar a dividir texto.

Lo que haremos será dividir la transcripción en trozos (chunks) de 1000 caracteres con una superposición de 30 caracteres, es decir, cada trozo tendrá 1000 caracteres y los 30 últimos caracteres de uno serán los mismos que los 30 primeros caracteres del siguiente trozo de texto.

In [37]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=30)

documentos = splitter.split_documents(documento_texto)

In [38]:
# número de trozos de texto desde la transcripción completa

len(documentos)   

165

In [39]:
# veamos los dos primeros

documentos[:2]

[Document(metadata={'source': 'txt/transcripcion.txt'}, page_content='Planned Chaos by Ludwig von Meises Introductory remarks The characteristic mark of this age of dictators, wars and revolutions is its anti-capitalistic bias. Most governments and political parties are eager to restrict the sphere of private initiative and free enterprise. It is an almost unchallenged dogma that capitalism is done for, and that the coming of all-round regimentation of economic activities is both inescapable and highly desirable. Nonetheless, capitalism is still very vigorous in the Western Hemisphere. Capitalist production has made very remarkable progress even in these last years. Methods of production were greatly improved. Consumers have been supplied with better and cheaper goods and with many new articles unheard of a short time ago. Many countries have expanded the size and improved the quality of their manufacturing. In spite of the anti-capitalistic policies of all governments and of almost al

### Encontrando los trozos relevantes

Una vez troceado el texto completo la pregunta surge naturalmente, ¿qué trozos deberíamos de darle al LLM como contexto?. No usaremos cualquiera de los 165 trozos de texto que hemos generado, sino que tenemos que usar aquellos que sean los más relevantes para la consulta que estamos haciendo. Es aquí donde entra el concepto de embeddings. Los embeddings son representaciones vectoriales de palabras, frases, u otros elementos en un espacio vectorial continuo de baja dimensión. Son una herramienta fundamental en el campo del procesamiento del lenguaje, ya que permiten que los modelos trabajen con información textual de una manera matemática. Recordemos que las máquinas solo entienden de números.


![chain5](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/chain5.png)


Usaremos el modelo de embeddings de OpenAI que ya viene incorporado en LangChain. Aquí también tendremos que usar la API KEY de OpenAI para el modelo.

In [40]:
from langchain_openai.embeddings import OpenAIEmbeddings


vectorizador = OpenAIEmbeddings()

Hagamos un ejemplo sencillo para entender el proceso antes de vectorizar todos los trozos de texto.

In [41]:
# vectorizamos una frase

consulta = vectorizador.embed_query('Hola que tal, esto es una clase de IA.')

In [42]:
# longitud del vector

print(f'Longitud del vector consulta: {len(consulta)}\n')


# primero 5 elementos del vector

print(consulta[:5])

Longitud del vector consulta: 1536

[-0.009476827457547188, -0.005400040186941624, -0.009671423584222794, -0.017513643950223923, 0.0057730162516236305]


La elección de 1536 como longitud de los embeddings en algunos modelos no es arbitraria, sino que se basa en una serie de consideraciones técnicas y de rendimiento. 

1. Capacidad de Representación: Un tamaño de vector de 1536 permite una rica representación de los datos. Con más dimensiones, el modelo puede capturar más detalles y matices en los datos. Esto es particularmente útil donde las sutilezas del lenguaje y el contexto pueden ser críticos. Un tamaño mayor puede ayudar a capturar relaciones más complejas y patrones más finos.

2. Rendimiento del Modelo: Aumentar el tamaño del embedding puede mejorar la capacidad del modelo para aprender y generalizar, hasta cierto punto. Modelos como BERT y sus variantes utilizan grandes vectores de embeddings para capturar un contexto lingüístico amplio y complejo, lo que les permite realizar bien en una variedad de tareas.

3. Equilibrio entre Rendimiento y Recursos: Elegir el tamaño del embedding también es un balance entre el rendimiento del modelo y el consumo de recursos computacionales, como memoria y tiempo de procesamiento. Un tamaño de 1536 puede ser un buen compromiso para obtener buenos resultados mientras se mantienen los requisitos de recursos en niveles manejables.

4. Compatibilidad y Estandarización: En algunos casos, el tamaño de los embeddings puede ser elegido para ser compatible con arquitecturas de modelo existentes o para estandarizar el entrenamiento y la implementación de modelos en diferentes entornos y aplicaciones. Esto facilita el uso de técnicas como la transferencia de aprendizaje y la implementación de modelos en producción.

5. Resultados Empíricos: La decisión final sobre el tamaño del vector de embeddings a menudo se basa en pruebas y experiencias empíricas. Los investigadores y desarrolladores prueban diferentes tamaños y seleccionan el que proporciona el mejor rendimiento en las tareas específicas de interés, basándose en métricas como la precisión, la recuperación, y el F1-score.


In [43]:
# ahora vectoricemos dos frase más 

frase1 = vectorizador.embed_query('¿Hoy estuviste estudiando o no?')

frase2 = vectorizador.embed_query('Estamos en clase, hay que ponerse a estudiar.')

Ahora que tenemos dos frases vectorizadas, queremos saber cual se parece más a la consulta, es decir, cual de las dos frases es la más relevante. El método más habitual para hacer esto es la similitud del coseno dado que estamos trabajando con vectores. La similitud del coseno mide el coseno del ángulo entre dos vectores en un espacio vectorial. El valor de esta similitud varía entre -1 y 1, donde 1 indica que los dos vectores son idénticos en orientación (ángulo de 0 grados), 0 sugiere que los vectores son ortogonales (ángulo de 90 grados) y no comparten ninguna información, -1 indica que los vectores son diametralmente opuestos (ángulo de 180 grados). De esta manera obtenemos una métrica de similitud entre los vectores. Desde la librería `scikit-learn` podemos usar la similitud del coseno.

In [44]:
from sklearn.metrics.pairwise import cosine_similarity

similitud_frase1 = cosine_similarity([consulta], [frase1])[0][0]

similitud_frase2 = cosine_similarity([consulta], [frase2])[0][0]

similitud_frase1, similitud_frase2

(0.7787063171053135, 0.8269535169772084)

Como vemos, la frase más parecida a nuestra consulta es la frase 2. Con esta estrategia podemos saber que trozos de texto son los más relevantes para la consulta específica que hagamos al LLM. El proceso será vectorizar todos los trozos de texto y luego calcular la similitud de los mismos con nuestra consulta para obtener los textos más relevantes y usarlos como contexto en el LLM. Para ello, lo adecuado es guardar los vectores obtenidos por el modelo de embedding en una base de datos y luego realizar una búsqueda de los vectores más relevantes.


![chain6](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/chain6.png)


## 9. Bases de datos vectoriales


LangChain provee de una buena cantidad de [bases de datos vectoriales](https://python.langchain.com/docs/integrations/vectorstores/) para el manejo de los embeddings. Vamos probar primero con una base de datos local en memoria, DocArrayInMemorySearch, que es un índice de documentos proporcionado por Docarray que almacena documentos en memoria. Es un excelente punto de partida para conjuntos de datos pequeños, donde es posible que no desees iniciar un servidor de base de datos.

In [45]:
from langchain_community.vectorstores import DocArrayInMemorySearch


# a docarray le pasamos los texto y el modelo de embedding

local_db = DocArrayInMemorySearch.from_documents(documentos, vectorizador)

/opt/anaconda3/envs/ia/lib/python3.9/site-packages/pydantic/_migration.py:283: UserWarning: `pydantic.error_wrappers:ValidationError` has been moved to `pydantic:ValidationError`.
  warnings.warn(f'`{import_path}` has been moved to `{new_location}`.')


In [46]:
# podemos extraer los k vectores más relevates con su similitud, donde k es el nº de trozos que queremos 

consulta = 'What are the key points of this book? Put them in a bullet point list.'

local_db.similarity_search_with_score(query=consulta, k=1)

[(Document(metadata={'source': 'txt/transcripcion.txt'}, page_content='They did not move the masses directly. It was the intellectuals who popularised them. The intellectual leaders of the peoples have produced and propagated the fallacies which are on the point of destroying liberty and western civilisation. The intellectuals alone are responsible for the mass slaughters which are the characteristic mark of our century. They alone can reverse the trend and pave the way for a resurrection of freedom. Not mythical material productive forces, but reason and ideas determine the course of human affairs. What is needed to stop the trend towards socialism and despotism is common sense and moral courage.'),
  0.7531981736270884)]

In [47]:
# también podemos extraer los k vectores más relevates sin similitud

local_db.similarity_search(query=consulta, k=1)

[Document(metadata={'source': 'txt/transcripcion.txt'}, page_content='They did not move the masses directly. It was the intellectuals who popularised them. The intellectual leaders of the peoples have produced and propagated the fallacies which are on the point of destroying liberty and western civilisation. The intellectuals alone are responsible for the mass slaughters which are the characteristic mark of our century. They alone can reverse the trend and pave the way for a resurrection of freedom. Not mythical material productive forces, but reason and ideas determine the course of human affairs. What is needed to stop the trend towards socialism and despotism is common sense and moral courage.')]

In [48]:
# o definir un recuperador de la base de datos para invocarlo que por defecto devuelve los 4 más relevantes

recuperador = local_db.as_retriever()

recuperador.invoke(consulta)

[Document(metadata={'source': 'txt/transcripcion.txt'}, page_content='They did not move the masses directly. It was the intellectuals who popularised them. The intellectual leaders of the peoples have produced and propagated the fallacies which are on the point of destroying liberty and western civilisation. The intellectuals alone are responsible for the mass slaughters which are the characteristic mark of our century. They alone can reverse the trend and pave the way for a resurrection of freedom. Not mythical material productive forces, but reason and ideas determine the course of human affairs. What is needed to stop the trend towards socialism and despotism is common sense and moral courage.'),
 Document(metadata={'source': 'txt/transcripcion.txt'}, page_content="a time-consuming procedure. They suggested a bold stroke. A small group of fanatics should be organised as the vanguard of the revolution. Strict discipline and unconditional obedience to the chief should make these profe

In [49]:
len(recuperador.invoke(consulta))

4

In [50]:
# aunque podemos definirlo para extraer los que queramos

# o definir un recuperador de la base de datos para invocarlo que por defecto devuelve los 4 más parecidos

recuperador = local_db.as_retriever(search_kwargs={'k': 5})

len(recuperador.invoke(consulta))

5

### Conectando la base de datos a la cadena


![chain7](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/chain7.png)


Ahora que tenemos la base de datos disponible en memoria, podemos conectarla a la cadena de LangChain, junto con la plantilla del prompt, el modelo LLM y el parser a string que habiamos definido anteriormente. Para ello vamos a usar dos objetos que nos proporciona LangChain.

RunnableParallel es esencialmente un diccionario cuyos valores son ejecutables, u objetos que pueden convertirse en ejecutables, como funciones. Ejecuta todos sus valores en paralelo, y cada valor se llama con la entrada general del RunnableParallel. El valor de retorno final es un diccionario con los resultados de cada valor bajo su clave correspondiente. Es útil para paralelizar operaciones, pero también puede ser útil para manipular la salida de un Runnable (protocolo ejecutable) para que coincida con el formato de entrada del siguiente Runnable en una secuencia. Se espera que la entrada al prompt sea un diccionario con las claves "contexto" y "pregunta". La entrada del usuario es solo la pregunta. Por lo tanto, necesitamos obtener el contexto usando nuestro recuperador y pasar la entrada del usuario.

RunnablePassthrough por sí solo permite pasar las entradas sin cambios. Esto típicamente se usa en conjunto con RunnableParallel para pasar datos a una nueva clave en el diccionario que entra en la cadena.

In [51]:
from langchain_core.runnables import RunnableParallel, RunnablePassthrough


# ejecutamos el RunnableParallel para obtener el contexto

contexto_recuperado = RunnableParallel(context=recuperador, question=RunnablePassthrough())

contexto_recuperado.invoke(consulta)

{'context': [Document(metadata={'source': 'txt/transcripcion.txt'}, page_content='They did not move the masses directly. It was the intellectuals who popularised them. The intellectual leaders of the peoples have produced and propagated the fallacies which are on the point of destroying liberty and western civilisation. The intellectuals alone are responsible for the mass slaughters which are the characteristic mark of our century. They alone can reverse the trend and pave the way for a resurrection of freedom. Not mythical material productive forces, but reason and ideas determine the course of human affairs. What is needed to stop the trend towards socialism and despotism is common sense and moral courage.'),
  Document(metadata={'source': 'txt/transcripcion.txt'}, page_content="a time-consuming procedure. They suggested a bold stroke. A small group of fanatics should be organised as the vanguard of the revolution. Strict discipline and unconditional obedience to the chief should mak

In [52]:
# creamos la cadena con el contexto recuperado desde la bade de datos

cadena = contexto_recuperado | prompt | modelo | parser

cadena.invoke(consulta)

'- Intellectual leaders popularized fallacies that are destroying liberty and western civilization\n- Intellectuals are responsible for mass slaughters in the 20th century\n- Reason and ideas, not material forces, determine human affairs\n- Common sense and moral courage are needed to stop the trend towards socialism and despotism\n- The book discusses the Bolsheviks and Mensheviks, who both aimed for socialism but differed in tactical methods\n- The book criticizes the central organization of the economy in Nazi Germany and the illusion of private ownership\n- The book challenges the anti-capitalistic bias and argues that capitalism is still vigorous in the Western Hemisphere'

Como hemos comentado, Docarray almacena documentos en memoria local. Pero para una aplicación real no nos sirve. Si queremos hacer un despliegue del modelo para una aplicación en producción necesitaremos una base de datos que podamos alojar en un servidor. Veamos un par de opciones open source.

###  Chroma 

![chroma](https://raw.githubusercontent.com/Hack-io-AI/ai_images/main/chroma.jpeg)


Otra opción que tenomos es [ChromaDB](https://www.trychroma.com/), que es una base de datos especializada en el almacenamiento y recuperación eficiente de información lingüística, incluyendo datos de texto, anotaciones semánticas y sintácticas. ChromaDB es particularmente útil para el almacenamiento y la gestión de grandes cantidades de datos de lenguaje natural, lo que permite a los desarrolladores aprovechar al máximo los avances en algoritmos de aprendizaje automático y análisis de texto. Veamos un ejemplo.

In [53]:
from langchain_community.vectorstores import Chroma

In [54]:
# guardado en disco

chroma_db = Chroma.from_documents(documentos, vectorizador, persist_directory='save_db')

docs = chroma_db.similarity_search(consulta)

In [55]:
docs[0]

Document(metadata={'source': 'txt/transcripcion.txt'}, page_content='They did not move the masses directly. It was the intellectuals who popularised them. The intellectual leaders of the peoples have produced and propagated the fallacies which are on the point of destroying liberty and western civilisation. The intellectuals alone are responsible for the mass slaughters which are the characteristic mark of our century. They alone can reverse the trend and pave the way for a resurrection of freedom. Not mythical material productive forces, but reason and ideas determine the course of human affairs. What is needed to stop the trend towards socialism and despotism is common sense and moral courage.')

In [56]:
# carga desde disco

chroma_db = Chroma(persist_directory='save_db', embedding_function=vectorizador)

docs = chroma_db.similarity_search(consulta)

docs[0]

/var/folders/sq/8cd72b4n08vdtd2xm8zrhvm80000gq/T/ipykernel_66181/680532756.py:3: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-chroma package and should be used instead. To use it run `pip install -U :class:`~langchain-chroma` and import as `from :class:`~langchain_chroma import Chroma``.
  chroma_db = Chroma(persist_directory='save_db', embedding_function=vectorizador)


Document(metadata={'source': 'txt/transcripcion.txt'}, page_content='They did not move the masses directly. It was the intellectuals who popularised them. The intellectual leaders of the peoples have produced and propagated the fallacies which are on the point of destroying liberty and western civilisation. The intellectuals alone are responsible for the mass slaughters which are the characteristic mark of our century. They alone can reverse the trend and pave the way for a resurrection of freedom. Not mythical material productive forces, but reason and ideas determine the course of human affairs. What is needed to stop the trend towards socialism and despotism is common sense and moral courage.')

In [57]:
# cadena con traducción y con la recuperacion de ChromaDB con 5 documentos

cadena =  prompt | modelo | parser

cadena_traducida = (
    {'answer': cadena, 'language': lambda x: x['language']} | prompt_traductor | modelo | parser
)

cadena_traducida.invoke({'context': chroma_db.similarity_search(consulta, k=5), 
                         'question': consulta,
                         'language': 'Spanish'}).split('\n')

['Los líderes intelectuales son responsables de popularizar falacias que están destruyendo la libertad y la civilización occidental. Los intelectuales también son responsables de matanzas en masa en el siglo XX. La necesidad de revertir la tendencia hacia el socialismo y el despotismo con sentido común y coraje moral. Comparación entre dos patrones organizativos: uno en el que el gobierno controla todos los aspectos de la producción y otro en el que la propiedad privada se mantiene nominalmente pero es controlada por el gobierno al estilo español.']

Desde aquí, ya podríamos hacer un despligue de nuestra cadena para implementarla en una aplicación en producción.

## 10. Resumen paso a paso

In [58]:
%%time

# librerias
from langchain_openai.chat_models import ChatOpenAI
from langchain.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_community.document_loaders import TextLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_openai.embeddings import OpenAIEmbeddings
from langchain_community.vectorstores import Chroma

from pytube import YouTube
import whisper
import tempfile

import os                           
from dotenv import load_dotenv      

load_dotenv()


# importamos la api key de OpenAI
OPENAI_API_KEY = os.getenv('OPENAI_API_KEY')


# iniciamos modelo gpt3.5 turbo
modelo = ChatOpenAI(openai_api_key=OPENAI_API_KEY, model='gpt-3.5-turbo')



# plantilla de texto con un contexto y una pregunta
plantilla = '''
            Answer the question based on the context below. If you can't 
            answer the question, reply "I don't know".

            Context: {context}

            Question: {question}
            '''

# carga de la plantilla en el prompt
prompt = ChatPromptTemplate.from_template(plantilla)


# traduccion
prompt_traductor = ChatPromptTemplate.from_template('Translate {answer} to {language}')


# extraccion de datos
# definimos la url del video
VIDEO_URL = 'https://www.youtube.com/watch?v=7EnHeZXLzTc'

# usamos pytube para extraer el video
youtube = YouTube(VIDEO_URL)
audio = youtube.streams.filter(only_audio=True).first()

# se carga el modelo base de Whisper 
modelo_whisper = whisper.load_model('base')


# ruta de guardado del archivo de texto
RUTA_TXT = 'txt/transcripcion.txt'

# si el archivo de texto no existe...
if not os.path.exists(RUTA_TXT):
    
    # abrimos el directorio temporal...
    with tempfile.TemporaryDirectory() as dir_temporal:
        
        # descargamos el audio de YouTube...
        archivo_audio = audio.download(output_path=dir_temporal)
        
        # y Whiper transcribe el audio a texto.
        transcripcion = modelo_whisper.transcribe(archivo_audio, fp16=False)['text'].strip()
        
        
        # se guarda el archivo de texto
        with open(RUTA_TXT, 'w') as archivo_texto:
            archivo_texto.write(transcripcion)


# cargamos el archivo de texto para usarlo en la cadena
with open(RUTA_TXT, 'r') as archivo_texto:
    transcripcion = archivo_texto.read()


# transformacion del dato
loader = TextLoader(RUTA_TXT)
documento_texto = loader.load()

splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=30)
documentos = splitter.split_documents(documento_texto)


vectorizador = OpenAIEmbeddings(api_key=OPENAI_API_KEY)



# guardado en disco, no sería necesario, hacer aparte
chroma_db = Chroma.from_documents(documentos, vectorizador, persist_directory='save_db')


# carga desde disco
chroma_db = Chroma(persist_directory='save_db', embedding_function=vectorizador)




# consulta
consulta = 'What are the key points of this book? Put them in a bullet point list.'


# parser a string
parser = StrOutputParser()




# cadena con traducción y con la recuperacion de ChromaDB con 5 documentos
cadena =  prompt | modelo | parser


cadena_traducida = (
    {'answer': cadena, 'language': lambda x: x['language']} | prompt_traductor | modelo | parser
)

cadena_traducida.invoke({'context': chroma_db.similarity_search(consulta, k=5), 
                         'question': consulta,
                         'language': 'Spanish'}).split('\n')

CPU times: user 1.37 s, sys: 188 ms, total: 1.55 s
Wall time: 11 s


['- Los intelectuales han popularizado falacias que están destruyendo la libertad y la civilización occidental',
 '- Los líderes intelectuales son responsables de las matanzas en masa en el siglo XX',
 '- La razón y las ideas, no las fuerzas productivas materiales, determinan los asuntos humanos',
 '- Se necesita sentido común y valor moral para detener la tendencia hacia el socialismo y el despotismo',
 '- La era de los dictadores, las guerras y las revoluciones tiene un sesgo anticapitalista',
 '- Muchos gobiernos y partidos políticos están ansiosos por restringir la iniciativa privada y la libre empresa',
 '- El capitalismo sigue siendo vigoroso en el Hemisferio Occidental a pesar de las políticas anticapitalistas',
 '- La producción capitalista ha hecho un progreso notable con métodos mejorados y suministro de bienes a España.']